### PR TITLE
Codependence module update

### LINKED ISSUE #428: Is this an error in the computing weights in HERC
In the Step3:
$$\alpha_1 = \frac{RC_1} {RC_1 + RC2}$$

the correct form may be:
$$\alpha_1 = 1 - \frac{RC_1} {RC_1 + RC2}$$

### PR DESCRIPTION
**Pull request by PanPip**
# Description

### Mutual Information estimators

This PR adds alternative Mutual Information estimators to the Codependence module:
- Standard Mutual Information estimator applied to the copula and its uniform margins
- Estimation of Mutual Information as (-1) * entropy of the copula

The implementation is based on the recent blog post by Dr. Gautier Marti [available here](https://gmarti.gitlab.io/qfin/2020/07/01/mutual-information-is-copula-entropy.html).

### Optimal Transport distance

This PR adds Optimal Transport distance measure to the Codependence module. Optimal Transport is well described in the blog post by Dr. Gautier Marti [available here](https://gmarti.gitlab.io/qfin/2020/06/25/copula-optimal-transport-dependence.html) and the [following research paper](https://arxiv.org/pdf/1610.09659.pdf).

The supported target copulas are:
-  A comonotone copula.
-  A countermonotone copula.
- A Gaussian copula with a custom correlation coefficient.
- A copula of both positive and negative correlation.
- A copula with some elements having extreme variations, while those of others are relatively small, and conversely.
- A copula with elements being positively correlated for small variations but uncorrelated otherwise.

Both MI estimators and OT distance can be used both in the individual functions - `get_mutual_info` and `optimal_transport_distance` and in the general `get_dependence_matrix` function.

### Other

This PR fixes #428: a small bug in the HERC documentation.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

New unit tests were added to the following files:

- [x] test_codependence.py

**Test Configuration**:
* Windows 10
* PyCharm

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -11,3 +11,5 @@ scipy==1.4.1
 sphinx==2.2.1 # needed for docs
 sphinx-rtd-theme==0.4.3 # needed for docs
 statsmodels==0.11.1
+cython==0.29.21
```
Comment (Jackal08) [inline requirements.txt:14]: This is fine because it is for developers.

File: setup.cfg [hunk 2]
```diff
@@ -53,6 +53,8 @@ install_requires =
     statsmodels==0.11.1
     matplotlib==3.2.1
     cvxpy==1.1.1
+    cython==0.29.21
```
Comment (Jackal08) [inline setup.cfg:56]: I am not so sure about adding these 2 as hard dependencies. 

Ok, let's do it, we are going to solve this problem with the Docker containers.

Make double sure the unit tests that use POT are very robust! If there is a dependency problem on their side, I want our tests to pick up that the code broke.

File: docs/source/codependence/information_theory_metrics.rst [hunk 3]
```diff
@@ -75,6 +75,55 @@ It can also be normalized using a known upper boundary:
...
+
+    \begin{gather*}
+    X_{unif} = \frac{X_{ranked}}{N}\\
+
+    Y_{unif} = \frac{Y_{ranked}}{N}\\
+
+    I[X, Y] = (-1) * H[C(X, Y)]
+    \end{gather*}
+
+According to Dr. Marti, these two estimators have some advantages over the standard approach:
```
[the remark this answers is not in the document]
Comment (PanPip) [reply docs/source/codependence/information_theory_metrics.rst:113]: You're right, will do

File: docs/source/codependence/codependence_marti.rst [hunk 4]
```diff
@@ -1,4 +1,4 @@
-.. _implementations-codependence_marti:
+.. _codependence-codependence_marti:
```
[the remark this answers is not in the document]
Comment (PanPip) [reply docs/source/codependence/codependence_marti.rst:1]: Will do

File: docs/source/codependence/codependence_marti.rst [hunk 5]
```diff
@@ -177,7 +177,8 @@ The following example shows how the above functions can be used:
 .. code-block::
 
    import pandas as pd
-   from mlfinlab.codependence import spearmans_rho, gpr_distance, gnpr_distance
+   from mlfinlab.codependence import spearmans_rho, gpr_distance, gnpr_distance,
```
[the remark this answers is not in the document]
Comment (PanPip) [reply docs/source/codependence/codependence_marti.rst:180]: Done

File: docs/source/codependence/correlation_based_metrics.rst [hunk 6]
```diff
@@ -169,7 +169,9 @@ The following examples show how the described above correlation-based metrics ca
 
     import pandas as pd
     from mlfinlab.codependence import distance_correlation, angular_distance,
```
[the remark this answers is not in the document]
Comment (PanPip) [reply docs/source/codependence/correlation_based_metrics.rst:171]: Done, also to other examples in the Codependence module.

File: docs/source/codependence/optimal_transport.rst [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+.. figure:: images/optimal_transport_distance.png
+   :scale: 100 %
+   :align: center
+   :figclass: align-center
+   :alt: Optimal Transport distance
+
+   Dependence is measured as the relative distance from independence to the nearest target-dependence: comonotonicity or
+   countermonotonicity. `(Blog post by Gautier Marti) <https://gmarti.gitlab.io/qfin/2020/06/25/copula-optimal-transport-dependence.html>`__
+
+"With this geometric view,
```
[the remark this answers is not in the document]
Comment (PanPip) [reply docs/source/codependence/optimal_transport.rst:34]: Done

File: setup.cfg (hunk 2 above)
Comment (PanPip) [reply setup.cfg:56]: Ok, we can later transfer them to a soft dependency if needed (or if these methods won't be popular).

**Pull request merged**